slackscot.go

```diff
@@ -126,83 +142,179 @@ func init() {
 		},
-	})
+	}
 }
 
-func onHeardMessage(message *slack.Message) {
-	hearActions.handle(message, false)
+// Close closes slackscot (its plugins) gracefully
+func (s *Slackscot) Close() {
+	for _, p := range s.plugins {
+		p.Close()
+	}
 }
 
-func onAskedMessage(message *slack.Message) {
-	log.Printf("%s-> %s", message.From, message.Text)
+// Run starts the Slackscot and loops until the process is interrupted
+func (s *Slackscot) Run(config config.Configuration) (err error) {
+	api := slack.New(
+		config.Token,
+		slack.OptionDebug(config.Debug),
+		slack.OptionLog(log.New(os.Stdout, "slackscot: ", log.Lshortfile|log.LstdFlags)),
+	)
+
+	rtm := api.NewRTM()
+	go rtm.ManageConnection()
 
-	matches := tellCommand.FindStringSubmatch(message.Text)
-	if len(matches) > 0 {
-		message.Tell(matches[3], matches[1]+": "+matches[4])
-		return
+	// Initialize slackscot (and all its plugins)
+	err = s.init(config)
+	if err != nil {
+		return err
 	}
 
-	handled := commands.handle(message, true)
+	// Register the close of slackscot when the process terminates
+	defer s.Close()
 
-	if !handled {
-		var response string
+	for msg := range rtm.IncomingEvents {
+		switch e := msg.Data.(type) {
+		case *slack.HelloEvent:
+			// Ignore hello
 
-		if defaultAction != nil {
-			// if sentence not matched to a supported request then fallback to a default
-			response = defaultAction(message)
-		} else {
-			response = "Je comprends pas.\n_Dis-moi_ `help` _pour avoir la liste des commandes_"
-		}
+		case *slack.ConnectedEvent:
+			log.Println("Infos:", e.Info)
+			log.Println("Connection counter:", e.ConnectionCount)
+
+		case *slack.MessageEvent:
+			s.processMessageEvent(rtm, e)
+
+		case *slack.PresenceChangeEvent:
+			log.Printf("Presence Change: %v\n", e)
+
+		case *slack.LatencyReport:
+			log.Printf("Current latency: %v\n", e.Value)
+
+		case *slack.RTMError:
+			log.Printf("Error: %s\n", e.Error())
 
-		log.Printf("me-> %s", response)
-		if err := message.Respond(response); err != nil {
-			// gulp!
-			log.Printf("Error responding to message: %s\nwith Message: '%s'", err, response)
+		case *slack.InvalidAuthEvent:
+			log.Printf("Invalid credentials")
+			return
+
+		default:
+			// Ignore other events
 		}
 	}
+
+	return nil
 }
 
-func Run(slackscot Slackscot, config config.Configuration) (err error) {
-	conn, err := slack.Connect(config.Token)
-	if err != nil {
-		log.Fatal(err)
+// processMessageEvent handles high-level processing of all slack message events.
+//
+// TODO (coming soon):
+// In the case of updated messages or deleted messages,
+// slackscot will look for an existing reply to that message. In the case of a deleted message that triggered a response from the bot,
+// that message will be deleted from slack. For a changed message, slackscot will either delete the message if not handled anymore or
+// will update the original response if it is still handled
+func (b *botEngine) processMessageEvent(c *slack.RTM, msgEvent *slack.MessageEvent) {
```

Comment:
I've got a little better version (not included here) with the other improvements I'm making. I'm having this `function`   handle retrieval of cached preexisting responses and then having the `routeMessage` (maybe I'll rename that one) return all the new/updated responses. That way, this function is the only one dealing with the actual reception/sending (but also the coming deletion and updating) of messages instead of having it scattered in a few places.